src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -125,12 +184,12 @@ public void StopProcessingNextRequest()
                     return;
                 }
 
-                if (!_stopping)
+                if (_stateManager.State != ConnectionState.Closed)
                 {
                     await _frameWriter.WriteSettingsAsync(_serverSettings);
                 }
 
-                while (!_stopping)
+                while (_stateManager.State == ConnectionState.Open || (_stateManager.State == ConnectionState.Closing) && !_streams.IsEmpty)
```

Comment:
Nit: add an extra parenthesis around the `&&` part of the condition for clarity.